hop/io.py

```diff
@@ -249,39 +270,77 @@ def unpack(message, metadata=False):
         else:
             return payload
 
+    def mark_done(self, metadata):
+        """Mark a message as fully-processed.
+
+        Args:
+            metadata: A Metadata instance containing broker-specific metadata.
+
+        """
+        tp = confluent_kafka.TopicPartition(
+            topic=metadata.topic,
+            partition=metadata.partition,
+            offset=metadata.offset + 1
+        )
+        # access Kafka consumer to store offsets
+        self._consumer._consumer.store_offsets(offsets=[tp])
```

Comment:
This doesn't seem right. We shouldn't be reaching into a private variable (the inner `_consumer`) like this. I would expect you to use `self._consumer.mark_done` ([ref](https://github.com/astronomy-commons/adc-streaming/blob/master/adc/consumer.py#L82-L88)). Does that work?

Reply:
Yeah, I'm not too happy about that either. The workaround is due to `mark_done()` only allowing `confluent_kafka.Message` as an input which won't be passed directly to the user here and can't be instantiated either so I can't create one and pass it in. An alternative could be to allow `adc.consumer.Consumer.mark_done()` to take in `TopicPartition` as well.

Reply:
I propose including the raw message in the public `Metadata` class's attributes, then. It can be a private attribute, `_raw: confluent_kafka.Message`, for example. Would that work?

Reply:
That sounds like a good approach. I can do some finagling so that you don't have to have two copies of all the relevant properties via a `__post_init__` method and just reference the `Message` directly.

Reply:
Alright, that should be in now. Turns out `__post_init__` and `frozen=True` don't mix well for dataclasses without an ugly workaround, and adding the frozen property has a tiny performance hit (https://docs.python.org/3/library/dataclasses.html#frozen-instances) so I'm inclined to drop it.